core/trino-main/src/test/java/io/trino/sql/ir/optimizer/TestSimplifyRedundantCase.java

```diff
@@ -92,7 +111,54 @@ void test()
+                                        IrExpressions.not(PLANNER_CONTEXT.getMetadata(), new Comparison(IDENTICAL, new Reference(BOOLEAN, "z"), TRUE))))));
+
+        assertThat(optimize(
+                new Case(
+                        ImmutableList.of(
+                                new WhenClause(new Reference(BOOLEAN, "a1"), TRUE),
+                                new WhenClause(new Reference(BOOLEAN, "a2"), FALSE),
+                                new WhenClause(new Reference(BOOLEAN, "a3"), FALSE),
+                                new WhenClause(new Reference(BOOLEAN, "a4"), TRUE)),
+                        TRUE)))
+                .isEqualTo(Optional.of(IrUtils.or(
+                        new Comparison(IDENTICAL, new Reference(BOOLEAN, "a1"), TRUE),
+                        IrUtils.and(
+                                IrExpressions.not(PLANNER_CONTEXT.getMetadata(), new Comparison(IDENTICAL, new Reference(BOOLEAN, "a2"), TRUE)),
+                                IrExpressions.not(PLANNER_CONTEXT.getMetadata(), new Comparison(IDENTICAL, new Reference(BOOLEAN, "a3"), TRUE)),
+                                IrUtils.or(
+                                        new Comparison(IDENTICAL, new Reference(BOOLEAN, "a4"), TRUE),
+                                        TRUE)))));
+
+        assertThat(optimize(
+                new Case(
+                        ImmutableList.of(
+                                new WhenClause(new Reference(BOOLEAN, "a1"), FALSE),
+                                new WhenClause(new Reference(BOOLEAN, "a2"), FALSE),
+                                new WhenClause(new Reference(BOOLEAN, "a3"), TRUE),
+                                new WhenClause(new Reference(BOOLEAN, "a4"), FALSE)),
+                        TRUE)))
+                .isEqualTo(Optional.of(
+                        IrUtils.and(
+                                IrExpressions.not(PLANNER_CONTEXT.getMetadata(), new Comparison(IDENTICAL, new Reference(BOOLEAN, "a1"), TRUE)),
+                                IrExpressions.not(PLANNER_CONTEXT.getMetadata(), new Comparison(IDENTICAL, new Reference(BOOLEAN, "a2"), TRUE)),
+                                IrUtils.or(
+                                        new Comparison(IDENTICAL, new Reference(BOOLEAN, "a3"), TRUE),
+                                        IrExpressions.not(PLANNER_CONTEXT.getMetadata(), new Comparison(IDENTICAL, new Reference(BOOLEAN, "a4"), TRUE))))));
+
+        assertThat(optimize(
+                new Case(
+                        ImmutableList.of(
+                                new WhenClause(new Reference(BOOLEAN, "a1"), TRUE),
+                                new WhenClause(new Reference(BOOLEAN, "a2"), FALSE),
+                                new WhenClause(new Reference(BOOLEAN, "a3"), FALSE)),
+                        FALSE)))
```

Comment:
What if all the when clause return `false` and default return `false` -> we could convert them to `FALSE` right ? Would we handle them here or at different rule ?

Reply:
Good catch. There was supposed to be a test for that at the top of this method, but I realize I made a copy/paste mistake.

Reply:
Updated.